### ISSUE TITLE
XFEMSpace be complex?

### ISSUE THREAD

**Issue by Shaoqigit** [ngsxfem/ngsxfem#9]
Dear developer,

These days, I am working on your package to solve some complex-valued problems (H1 complex space).
I create an interface with levelset function as:
```
levelset = (x-0)
lsetp1 = GridFunction(H1(mesh,order=1))
InterpolateToP1(levelset,lsetp1)
```

then, create a solution space in complex H1
`Vh = H1(mesh, order=order, complex=True)`
After this, I have two choices, either use the cutFEM in the usual space as:
```
ci = CutInfo(mesh, lsetp1)
VhG = FESpace([Compress(Vh,GetDofsOfElements(Vh,ci.GetElementsOfType(cdt))) for cdt in [HASNEG,HASPOS]], complex=True)
```
which works well, or I would like to use XFEMSpace with the enrichment as:
```
ci = CutInfo(mesh, lsetp1)
Vhx = XFESpace(Vh, ci)
VhG = FESpace([Vh, Vhx], complex=True)
```
this construction poses an issue: **Compound space of spaces with complex and real spaces is not allowed**
I checked the Vhx with .is_complex, it returns me false. Ok, I can understand the error, but I did not find a way to make XFEMSpace in complex space, there is no argument in the constructor of XFEMSpace to assign it as complex space.

Was I missing some functionalities or this has not been considered in the code? look forward to your reply and excuse me all these basic questions.

Best,
Shaoqi

**Comment by schruste** [comment]
Hi Shaoqi,

Nice that are really digging into the problems already :). We haven't been working with complex problems that much so far with ngsxfem and hence don't meet these problems so often :innocent: . In your case, you'dexpect that the XFESpace is complex because Vh is complex. This is not implemented at the moment so that the error is thrown. I will add this soon. But to be sure, could you first try the following work-around: `Vhx = XFESpace(Vh, ci, flags={ "complex" : True })`?

Best,
Christoph

**Comment by Shaoqigit** [comment]
Yes, for the coherence, my approximation space is in complex value, the enrichment XFEM space do need in the complex as well.
I have tried the work-around and the compound error is gone,  but a new error occurs for the code (impose the Neumann BC with my exact flux function):
```
f = LinearForm(VhG)
f += flux[0] * v[0] * ds('left')
f += flux[1] * v[1] * ds('right')
```
error: **netgen.libngpy._meshing.NgException: Testfunction does not support BND-forms, maybe a Trace() operator is missing**

This code is good with CutFEM space and compute the solution correctly, so the problem is from the XFEMSpace?

Best,
Shaoqi

**Comment by schruste** [comment]
Oh, yes, the boundary evaluator is also not implemented for the XFESpace :see_no_evil: . 

I just added it and uploaded it to the master branch on github. please update your xfem installation (master branch) and try again. Also: For next time, please attach a small minimal (not) working example which would help us to figure out a solution.

Best,
Christoph

**Comment by Shaoqigit** [comment]
Sorry for this late reply, I just update my installation by pip3,  but these two errors are still here. I am not sure if I can update the newest version via pip3 install git+https://github.com/ngsxfem/ngsxfem.git@master

Best,
Shaoqi

**Comment by schruste** [comment]
The pip install is enough but you should add `--upgrade` to update it. If it still doesn't work, please attach a minimal example.

Best,
Christoph

**Comment by Shaoqigit** [comment]
It does not work, here is a minimal example
[test.zip](https://github.com/ngsxfem/ngsxfem/files/6445422/test.zip)
Thanks

**Comment by schruste** [comment]
Ah, I see. I will have a look at this. It may take some time until I can work on it though.

Best,
Christoph

**Comment by Shaoqigit** [comment]
It's ok, take your time.
Actually, I am a Ph.D student in computational acoustics, I mainly work on an in-house c++ code of our lab (It works quite wells ).
Just for personal interest, I found yours xfem package with python interface and all interactive trick on jupyter very attractive among "all" open source package on the site. 
I will keep trying your code for some acoustic problems during my space-time and keep in touch with your side!

Best, 
Shaoqi
